### PR TITLE
fix(security): Block environment-provisioned provider credentials from tenant-chosen endpoints

### PR DESCRIPTION
**Pull request by erichare**
## Summary

Fixes HackerOne report **H1-4000668**, CWE-522 Insufficiently Protected Credentials).

### Root cause

Model-provider components ship with an `api_key` field whose default value is the **name** of an environment variable, and expose a tenant-settable base URL. Two default-on mechanisms (`store_environment_variables` seeding and the `fallback_to_env_var` load-from-DB fallback) resolve that name to the operator's key held in the server process environment — a value the tenant may *use* but is deliberately barred from *reading* (outputs are scrubbed via `_sanitize_secret_values`). Nothing checked the other direction: the resolved key was placed in an `Authorization` header and sent to whatever host the tenant's base URL named. The existing connector SSRF guard (`provider_ssrf.py`) only blocks private/reserved ranges, so an attacker-owned **public** host was permitted — its own docstring acknowledged this gap. The leak fires on `GET /v1/models` during model construction/discovery, so a flow that errors out has already leaked.

### Fix

A new additive credential-egress control in `src/lfx/src/lfx/base/models/provider_ssrf.py`:

- `ensure_credential_endpoint_allowed(api_key, base_url, default_url=...)` refuses (ValueError) to forward a credential whose resolved value **exactly matches a value held in the server process environment** to any endpoint other than the provider's own default or an operator-allowlisted host. Value matching covers both delivery routes (seeded credential variables and the env fallback) with one check, and a tenant cannot arrange a false positive without already knowing the value they cannot read. Tenant-owned keys (typed or stored as their own global variables) are unaffected, so bring-your-own-key flows against custom endpoints keep working.
- New operator setting `provider_credential_allowed_hosts` (`LANGFLOW_PROVIDER_CREDENTIAL_ALLOWED_HOSTS`, exact host / `host:port` / `*.wildcard`) for deployments that legitimately front a provider with a proxy while provisioning the key via the environment.

Guarded components (the five named in the report, plus OpenAI embeddings which shares the field pattern): OpenAI chat model and embeddings (`lfx-openai`), xAI, LiteLLM Proxy, LM Studio, and NVIDIA (`lfx-bundles`) — including the NVIDIA/xAI model-discovery paths that send the key before any inference call.

### Testing

- `src/lfx/tests/unit/base/models/test_provider_ssrf.py`: 18 new tests for the guard (blocks env-sourced key to public custom host; allows provider default, empty URL, tenant-owned keys, allowlisted/wildcard hosts, secret wrappers, short low-entropy env values) — 51 passed.
- `src/bundles/openai/tests/test_openai_model_ssrf.py`: 5 new egress regression tests (chat + embeddings) — 34 passed, 2 skipped.
- `src/bundles/lfx-bundles/tests/test_nvidia_component.py`: 4 new egress regression tests incl. the model-discovery path — 25 passed (with existing connector SSRF suites).
- Settings composition tripwire updated for the new field — 41 passed.
- Manual wiring verification for xAI / LiteLLM Proxy / LM Studio (blocked env-sourced keys to attacker host; allowed tenant key to custom host and env key to default endpoint).
- `ruff check` / `ruff format` clean; `mypy` on `provider_ssrf.py` shows only the two pre-existing errors.

### Notes for reviewers

- Not fixable by SSRF configuration: this is intentionally a separate control from `ssrf_allowed_hosts` (which governs internal-address SSRF). The two settings are kept distinct so an operator widening SSRF allowances does not silently widen credential egress.
- The Anthropic component (`lfx-anthropic`, opt-in bundle) has the same field pattern but no env-var-name shipped default; the helper is available for it and other providers as follow-up if desired.

## Summary by CodeRabbit

* **Security**
  * Server-provisioned provider credentials are now prevented from being sent to tenant-selected endpoints unless the destination is the provider’s default or an approved host.
  * Added support for exact host, host-and-port, and wildcard host allowlists through security settings.
  * Tenant-provided credentials remain usable with custom endpoints under existing protections.

* **Updates**
  * Updated OpenAI and bundled provider package versions to include these safeguards.

### CODE REVIEW

File: src/lfx/src/lfx/base/models/provider_ssrf.py
Comment (Adam-Aghili) [inline src/lfx/src/lfx/base/models/provider_ssrf.py:1]: **Security follow-up: check the SDK’s effective credential before allowing an empty key**
We can address this in a follow-up PR if needed. 

At `ensure_credential_endpoint_allowed` (around line 214), an empty key returns before the endpoint policy runs. OpenAI chat converts an empty input to `api_key=None`, after which the SDK loads `OPENAI_API_KEY` from the server environment. A tenant-selected public base URL can therefore receive the operator’s credential despite this guard.

I confirmed the guard accepts `None` and the installed ChatOpenAI client resolves a synthetic environment key while retaining the custom destination. No outbound request was made. This remains a substantial credential-disclosure gap; the follow-up should land before the security issue is considered closed.

**AI prompt for the follow-up PR**

```text
Fix the empty-key SDK fallback bypass in PR #15211. Resolve the effective provider credential before applying ensure_credential_endpoint_allowed, or reject missing credentials on custom endpoints where safe resolution is unavailable. Pass the same validated credential explicitly to the SDK so later environment fallback cannot change it. Audit chat, embeddings, and the other guarded provider call sites for equivalent fallback behavior. Preserve tenant-owned keys, provider defaults, and the operator credential-host allowlist. Add regression tests using a synthetic OPENAI_API_KEY and an empty/None component key: reject an unapproved custom endpoint before any request; allow the provider default and approved hosts. Exercise the real SDK fallback with network calls intercepted, and run the relevant provider tests and Ruff. Do not use real credentials or external endpoints in tests.
```

**Review by Adam-Aghili — CHANGES_REQUESTED** [review]
1 issue. prefer it in this PR but could be follow up

File: src/lfx/src/lfx/base/models/provider_ssrf.py
Comment (erichare) [reply src/lfx/src/lfx/base/models/provider_ssrf.py:1]: Fixed in c5767f4a94 — landed here rather than as a follow-up.

**Confirmed, and reproduced on both OpenAI call sites.** `ChatOpenAI(api_key=None, base_url=<tenant>)` resolves `OPENAI_API_KEY` out of the server environment and keeps the tenant's destination; `OpenAIEmbeddings` does the same via its async client:

```
ChatOpenAI  api_key=None -> auth header carries $OPENAI_API_KEY: True
            base_url                                           : https://attacker.example.test/v1
Embeddings  api_key=None -> auth header carries $OPENAI_API_KEY: True
            base_url                                           : https://attacker.example.test/v1/
```

One detail that shaped the fix: `api_key=""` behaves differently from `None` — the OpenAI SDK raises `OpenAIError: Missing credentials` rather than substituting an environment key. So the resolvable case is specifically `None`, and lumping "empty" into one bucket would over-block a component that is deliberately sending no credential.

**What changed.** `ensure_credential_endpoint_allowed` gained `sdk_env_fallback` — the variable(s) the SDK reads when the component passes nothing — and now judges the credential that will actually be sent:

```python
ensure_credential_endpoint_allowed(
    api_key_value, self.openai_api_base,
    default_url=DEFAULT_OPENAI_API_BASE,
    sdk_env_fallback="OPENAI_API_KEY",
)
```

Each call site passes the *same expression its SDK constructor receives*, so the guard and the SDK cannot disagree about what the credential is. Audited all six: OpenAI chat, OpenAI embeddings, xAI, LM Studio, LiteLLM Proxy, NVIDIA. Worth flagging — xAI, LM Studio and LiteLLM Proxy all build a `ChatOpenAI`, so their fallback variable is `OPENAI_API_KEY`, not the provider's own. An empty xAI key would have leaked the operator's *OpenAI* key.

**Tests.** Component-level regressions with a synthetic `OPENAI_API_KEY` and an empty/None key: rejected before any client is constructed, provider default and allowlisted hosts still allowed. Plus one that exercises the real SDK unmocked (no request is made — the client is only constructed and inspected) so the guard's premise is pinned: if langchain-openai ever stops resolving the environment key, that test fails instead of the guard silently over-blocking. Mutation-verified — neutering the new branch fails 5 of the new tests and no others.

Ran: 8332 lfx unit / 42 openai bundle / 219 lfx-bundles, all green.

File: src/lfx/src/lfx/base/models/provider_ssrf.py [hunk 1]
```diff
@@ -118,3 +129,102 @@ def openai_compatible_client_kwargs(base_url: str | None, *, default_url: str |
...
+    regardless of which route delivered it. A tenant cannot arrange a false positive
+    without already knowing the value they are not allowed to read.
+
+    Accepts plain strings and ``SecretStr``-style wrappers. Values shorter than
+    ``_MIN_ENV_CREDENTIAL_LENGTH`` never match.
+    """
+    text = secret_value_to_str(value)
+    if not text or len(text) < _MIN_ENV_CREDENTIAL_LENGTH:
+        return False
+    return any(text == env_value for env_value in os.environ.values() if len(env_value) >= _MIN_ENV_CREDENTIAL_LENGTH)
```
[the remark this answers is not in the document]
Comment (erichare) [reply src/lfx/src/lfx/base/models/provider_ssrf.py]: Valid — fixed in c5767f4a94. The length floor is gone; `is_env_sourced_credential` now matches any non-empty environment value.

```python
text = secret_value_to_str(value)
if not text:
    return False
return any(text == env_value for env_value in os.environ.values())
```

The asymmetry decides it: a false negative hands the operator's key to a tenant-chosen endpoint, a false positive refuses one custom endpoint with an actionable message. Confirmed your reachability claim too — neither the LiteLLM path nor the local-provider path imposes a length floor of its own, so a short virtual key or NIM token was failing open through exactly those sinks.

Two tests flipped to match: `test_should_block_short_env_sourced_keys` and `test_should_match_short_values`, each carrying the rationale so the old heuristic isn't reintroduced as a "fix" later.

File: src/bundles/openai/src/lfx_openai/components/openai/openai.py [hunk 2]
```diff
@@ -75,6 +75,13 @@ class OpenAIEmbeddingsComponent(LCEmbeddingsModel):
     ]
 
     def build_embeddings(self) -> Embeddings:
+        # A key that resolves to a value held in the server process environment is the
+        # operator's credential: it may only leave the deployment to an endpoint the
+        # operator sanctioned (the default OpenAI endpoint or an allowlisted host).
+        ensure_credential_endpoint_allowed(
+            self.openai_api_key, self.openai_api_base, default_url=DEFAULT_OPENAI_API_BASE
+        )
```
[the remark this answers is not in the document]
Comment (erichare) [reply src/bundles/openai/src/lfx_openai/components/openai/openai.py:89]: Valid — fixed in c5767f4a94, with one deliberate deviation.

Allowlisting a host sanctions the *destination*; it does not sanction putting the key on the wire in cleartext, and an allowlist entry has no way to express a scheme. So an allowlisted host is now held to HTTPS:

```python
if entry is None:
    raise ValueError(...)                      # host not sanctioned at all
if parsed.scheme.lower() == "http" and not _parse_allowlist_entry(entry)[1]:
    raise ValueError(... "over cleartext HTTP" ...)
```

**The deviation:** rather than rejecting non-HTTPS unconditionally, an operator can opt in per host by writing the entry with an explicit scheme — `http://llm-gateway.internal:8000`. Self-hosted vLLM/NIM gateways on an internal network are legitimately plain HTTP, and with no escape hatch an operator who deliberately allowlisted such a host would have had no recourse. Since `LANGFLOW_PROVIDER_CREDENTIAL_ALLOWED_HOSTS` is new in this PR, nobody is relying on the looser behavior; HTTPS is the default and cleartext is now a conscious act.

Your ordering note was right and is preserved: the provider-default check runs first, so LM Studio's `http://localhost:1234/v1` and the LiteLLM Proxy default are untouched — covered by `test_should_not_apply_cleartext_rule_to_the_provider_default`.

This also caught a real consequence in-tree: the e2e harness allowlists its loopback OpenAI stub, which only serves plain HTTP, so `playwright.config.ts` now reads `http://127.0.0.1`. That the escape hatch was needed on the first real config is decent evidence it belongs.

File: src/bundles/lfx-bundles/src/lfx_bundles/nvidia/nvidia.py [hunk 3]
```diff
@@ -76,13 +79,16 @@ class NVIDIAModelComponent(LCModelComponent):
     ]
 
     def get_models(self, *, tool_model_enabled: bool | None = None) -> list[str]:
+        # Note: don't include the previous model, as it may not exist in available models from the new base url
+        # The key sent below may resolve to the operator's environment-provisioned
+        # credential; refuse to forward it to a tenant-chosen endpoint.
+        ensure_credential_endpoint_allowed(self.api_key, self.base_url, default_url=NVIDIA_DEFAULT_BASE_URL)
```
[the remark this answers is not in the document]
Comment (erichare) [reply src/bundles/lfx-bundles/src/lfx_bundles/nvidia/nvidia.py]: Valid — fixed in c5767f4a94. Your analysis of the constraints held up: `_NVIDIAClient.__init__` overwrites `get_session_fn` with `_create_session`, and `ChatNVIDIA` does not forward arbitrary kwargs, so there is no constructor-level fix.

NVIDIA is genuinely the outlier here — every other guarded provider takes `openai_compatible_client_kwargs`, which hands it redirect-free httpx clients. ChatNVIDIA owns a `requests.Session`, so there is no `http_client` kwarg to inject. And `Session.rebuild_auth` compares *hostname only*, so `https://allowed-host/v1` → `http://allowed-host:9999/` keeps the Authorization header and downgrades to cleartext in one hop.

The fix wraps the client's session factory rather than replacing it, so `verify_ssl`, adapters and proxies survive:

```python
def guarded(*args, _factory=factory, **kwargs):
    return refuse_redirects(_factory(*args, **kwargs))
```

`refuse_redirects` (new, in `lfx.utils.ssrf_requests`) overrides `resolve_redirects` to raise `SSRFProtectionError`. Refusing outright rather than re-validating each hop: an inference API answering `/chat/completions` with a redirect is not a flow any provider supports, and it avoids reimplementing DNS pinning on a transport we do not own. The `yield_requests=True` probe that `Session.send` makes even when not following redirects is exempted — that probe is not egress.

Applied to both `get_models` and `build_model`, and both `_client` and `_async_client` when present. Since the hook is a private attribute of a pinned SDK, a version that moves it logs a warning rather than failing open silently — the endpoint guard and the SSRF denylist still cover the initial request either way.

Tests cover all five redirect statuses, the same-host port change specifically, a normal response passing through, the `yield_requests` probe, session config surviving, and the fail-open warning firing.

**Comment by erichare** [comment]
All four review items addressed in c5767f4a94 — including @Adam-Aghili's, which is landed here rather than deferred to a follow-up since it was the remaining hole in the control this PR exists to add.

| Item | From | Status |
|---|---|---|
| Empty key → SDK resolves the operator's env credential | @Adam-Aghili | Fixed — `sdk_env_fallback` |
| NVIDIA transport follows redirects | CodeRabbit | Fixed — `refuse_redirects` on both paths |
| Cleartext HTTP to an allowlisted host | CodeRabbit | Fixed, with a per-host `http://` opt-in |
| 8-char floor exempted short env credentials | CodeRabbit | Fixed — floor removed |
| `host:port` allowlist coverage | CodeRabbit (nit) | Tests added |

**The through-line.** Three of these were the same mistake in different clothes: the guard was judging *the value the component happened to pass* instead of *what would actually go on the wire*. An absent key looked like no credential but became `OPENAI_API_KEY`. A short key looked like a flag value but was a real virtual key. An allowlisted host looked sanctioned but the scheme was never checked. The guard now resolves the effective credential and the effective transport before deciding.

**One finding worth surfacing beyond its thread:** xAI, LM Studio and LiteLLM Proxy all build a `ChatOpenAI`. Their SDK fallback variable is `OPENAI_API_KEY`, not the provider's own — so an empty xAI key pointed at a custom host would have leaked the operator's *OpenAI* credential. That cross-provider path is covered now.

**Behavior changes an operator could notice**, all narrow and all failing closed:
- A blank API key on a custom, non-allowlisted endpoint is now refused when the SDK's fallback variable is set. Previously it silently sent the operator's key.
- An allowlisted host must be HTTPS unless the entry carries an explicit `http://`. Provider defaults are exempt, so LM Studio and LiteLLM Proxy keep their http endpoints. This one had a real in-tree consequence: the e2e harness's loopback OpenAI stub only serves plain HTTP, so `playwright.config.ts` now reads `http://127.0.0.1`.
- A short credential matching an environment value is now blocked rather than exempted.

**Verification.** 8332 lfx unit, 42 openai bundle, 219 lfx-bundles — green. Ruff clean. The absent-key regressions are mutation-verified: neutering the new branch fails exactly the 5 new tests and nothing else. One test builds the real unmocked `ChatOpenAI` (no request issued — the client is constructed and inspected) so if langchain-openai ever stops resolving the environment key, that assumption fails loudly instead of the guard quietly over-blocking.

**Comment by erichare** [comment]
Also fixed the red **Test CI release scripts** check in 97290ee6b0 — it turned out to be pre-existing, not from the review fixes.

The version bump in 7079aabcd4 moved `src/bundles/openai` to 0.1.5 and pointed the root `pyproject.toml` at `>=0.1.5`, but `scripts/ci/bundle_profiles.json` was left declaring `>=0.1.4`, so `validate_contract` has been failing on every push since:

```
- enterprise-hardened: lfx-openai range differs from pyproject.toml: >=0.1.4,<1.0.0 != <1.0.0,>=0.1.5
```

Bumped the contract range and regenerated the lock with `manage_bundle_profiles.py compile` rather than hand-editing it. The diff is minimal — lfx-openai repinned 0.1.4 → 0.1.5 plus the profile digest restamp. `manage_bundle_profiles.py check` is clean and all 157 `scripts/ci` tests pass.

File: src/bundles/lfx-bundles/src/lfx_bundles/nvidia/nvidia.py [hunk 4]
```diff
@@ -76,14 +117,25 @@ class NVIDIAModelComponent(LCModelComponent):
...
         try:
             from langchain_nvidia_ai_endpoints import ChatNVIDIA
         except ImportError as e:
             msg = "Please install langchain-nvidia-ai-endpoints to use the NVIDIA model."
             raise ImportError(msg) from e
 
-        # Note: don't include the previous model, as it may not exist in available models from the new base url
-        model = ChatNVIDIA(base_url=self.base_url, api_key=self.api_key)
+        model = ChatNVIDIA(base_url=self.base_url, api_key=self.api_key or None)
+        _block_redirects(model)
```
Comment (Adam-Aghili) [inline src/bundles/lfx-bundles/src/lfx_bundles/nvidia/nvidia.py]: **Install redirect protection before discovery and cover the async transport**

Rechecked `97290ee6b0` with the locked `langchain-nvidia-ai-endpoints==1.0.4`. `_block_redirects` is too late for `get_models()`: `ChatNVIDIA(base_url=..., api_key=...)` calls `_NVIDIAClient.available_models` during construction when no model is supplied. A synthetic adapter confirmed that its initial `/v1/models` request followed a 307 to another host before this line ran.

Async inference also remains unprotected: `ChatNVIDIA._agenerate`/`_astream` call the same `_client` through `get_async_session_fn` (`aiohttp`), while this helper only replaces `get_session_fn` (`requests`). After applying the actual helper, a loopback-only 307 test still delivered the synthetic prompt body to a second, unapproved port. The authorization header was stripped in that test; the demonstrated impact is an unvalidated redirected request and prompt-body disclosure, not a proven key leak.

**AI implementation prompt**

Enforce the redirect policy before the NVIDIA SDK can make its first request, including constructor-time model discovery, and on both synchronous and asynchronous inference/streaming sessions. Avoid relying on a post-construction sync-only hook. Preserve TLS/proxy configuration and fail closed if the required transport hook cannot be installed. Add tests against the pinned real SDK with controlled adapters or loopback servers for discovery, async inference, and streaming; assert that no redirect destination is contacted and no prompt body is forwarded. Keep ordinary successful requests working and rerun the provider and transport suites.

**Review by Adam-Aghili — CHANGES_REQUESTED** [review]
Please cover constructor-time discovery and asynchronous NVIDIA transports with the redirect policy, as described in the review comment.

File: src/bundles/lfx-bundles/src/lfx_bundles/nvidia/nvidia.py (hunk 4 above)
Comment (erichare) [reply src/bundles/lfx-bundles/src/lfx_bundles/nvidia/nvidia.py]: Both confirmed and fixed in 0b13044224. Reproduced each one against the pinned `langchain-nvidia-ai-endpoints==1.0.4` with loopback servers before changing anything:

```
== discovery (constructor) ==            == async inference ==
  constructed, model: sink-model           completed
  sink contacted:                          sink contacted:
    {'path': '/v1/models'}                   {'path': '/v1/chat/completions',
                                              'body': '{"messages": [{"role": "user",
                                                       "content": "SECRET_PROMPT_CANARY"}], ...}'}
```

`ChatNVIDIA.__init__` resolves `available_models` whenever no model name is given, so `/v1/models` goes out from inside the constructor — the old hook could not have run yet. And `_agenerate`/`_astream` go through `get_async_session_fn` (aiohttp), which the `requests`-only hook never touched; the prompt body reached the second port intact.

**What changed.** `_install_redirect_policy()` now hardens `_NVIDIAClient` itself and is called *before* construction, at both call sites. Two seams, because the SDK uses two:

- `_create_session` / `_create_async_session` are wrapped at class level, covering every request issued after construction — sync, async and streaming.
- `__init__` is wrapped to pass those same wrapped creators as the `get_session_fn` / `get_async_session_fn` field values. The SDK only assigns them on its *last* line, so until then the fields still hold the plain `requests.Session` / `aiohttp.ClientSession` defaults — exactly what constructor-time discovery used. Passing the SDK's own bound creators keeps its `verify_ssl` and connector setup; the only difference from what it assigns later is the redirect refusal.

`refuse_aiohttp_redirects` is the aiohttp counterpart of `refuse_redirects`: every request is issued with `allow_redirects=False` *and* a redirect response is refused rather than returned, so a caller cannot resume the hop by reading `Location` itself. aiohttp is duck-typed, not imported, so lfx gains no dependency.

**Fail-closed, as asked.** The old helper logged a warning and continued when the SDK moved its hook. It now raises — continuing means a client that can follow a redirect with the operator's key attached.

**Tests** run the real pinned SDK against loopback servers: constructor discovery, sync inference, async inference, streaming — each asserting the redirect target received nothing — plus a happy path proving discovery, `invoke`, `ainvoke` and `astream` all still work normally, and two fail-closed cases. The four mock-SDK tests they replace could not have caught either gap: the fake had no constructor-time request and no async transport.

Mutation-verified: dropping the constructor seeding fails only `test_constructor_discovery_refuses_redirect`; dropping the aiohttp wrapper fails only the async and streaming tests.

**On the residual you flagged earlier.** This stops redirects; it still is not connection-time IP pinning, and I left that note in `get_models` with the wording corrected — the SDK's `_create_session` is now hooked, but hooking it does not let us dial a pre-resolved IP while keeping TLS SNI.

**Merge.** `release-1.12.3` is merged in (`cc00161811`). #15208 landed `validate_provider_base_url` and renamed the constant, so NVIDIA now runs both host checks — SSRF/HTTPS validation, then the credential-egress guard, then the redirect policy, then construction. Bundle bumped to 1.1.30 on top of upstream's 1.1.29 and the profile lock regenerated (`manage_bundle_profiles.py check` passes). Two NVIDIA credential-egress tests needed the SSRF host lookup neutralized — the canary domain has no DNS record, which upstream's new check now resolves — using the same seam the shared-helper tests already patch.

Green locally: 8597 lfx unit, 253 lfx-bundles, 42 lfx-openai.

**Review by Adam-Aghili — APPROVED** [review]
Re-reviewed cc00161811d397fd4c33f09af09f37e04a7e1723. The redirect policy is installed before NVIDIA construction and now covers synchronous, asynchronous and streaming sessions. The SDK-hook checks fail closed. This addresses my outstanding change request.

Validated 185 tests against the reviewed module snapshots, including the pinned NVIDIA SDK 1.0.4 regression tests and ordinary-request checks; all passed. Ruff lint and format checks passed on all 14 changed Python files. Validation used local Python 3.12 dependencies; full CI remains required.
